### PR TITLE
Add some guidelines on tag use

### PR DESCRIPTION
**Pull request by inducer**

### CODE REVIEW

**Review by kaushikcfd — APPROVED** [review]
I generally agree with the tone here. However, I also wanted to shed light on the discussion in
#55. Wrt #55, the text is completely wrong.
1. Stripping off send-tags would be illegal.
2. And in the definition of "semantics" agreed upon, "sends" are a part of the semantic information.

**Comment by inducer** [comment]
> Stripping off send-tags would be illegal.

@matthiasdiener pointed this out as well, in my discussion with him earlier today, and it I agree. (That discussion is, in part, what prompted this PR.) The problem is that I kind of like the way that [the current direction in #55](https://github.com/inducer/pytato/issues/55#issuecomment-808592686) allows the sends to be "along for the ride." What we could do is create an extra array attribute that would allow other nodes to "come along". `tag_along_nodes`? This seems awkward. Got any better ideas?

File: doc/design.rst [hunk 1]
```diff
@@ -151,6 +151,27 @@ Reserved Identifiers
...
+Tags
+----
+
+In order to convey information about the computation from DAG construction
+time to processing/transformation/code generation time, each :class:`pytato.Array`
+node may be tagged (via the :attr:`pytato.Array.tags` attribute) with an arbitrary
+number of informational "tags". A tag is any subclass of :class:`pytools.tag.Tag`.
+Guidelines for tag use:
+
+- Tags *must not* carry semantic information; i.e. a computation must have the same
```
Comment (matthiasdiener) [inline doc/design.rst:163]: Any particular reason why it needs to be *must not* and not *should not* here? Imho, there are legitimate reasons for conveying semantic information, including the distributedsend stuff as well as tagging operations as unsuitable for GPU, for example (say, file I/O).

**Comment by kaushikcfd** [comment]
> What we could do is create an extra array attribute that would allow other nodes to "come along". tag_along_nodes

Agree with this. From a user's POV I don't see a big difference with prescribing the dependency via a tag or via an attribute.

File: doc/design.rst (hunk 1 above)
Comment (kaushikcfd) [reply doc/design.rst:163]: >  for example (say, file I/O).

Legitimate point, but I think the chances of allowing I/O into our DAG is slim.

Comment (inducer) [reply doc/design.rst:163]: IMO *must not* would provide the clearest delineation of what can and cannot go into a tag.

My point with this PR was precisely to provide a spot for this discussion.

**Comment by inducer** [comment]
I propose that we move forward with this, and we find another way (in #55) to bring the sends "along for the ride". @matthiasdiener OK with you?

Also, unrelated: CI just seems to hang. I'm a little confused by that.

**Review by matthiasdiener — APPROVED** [review]

**Pull request merged**